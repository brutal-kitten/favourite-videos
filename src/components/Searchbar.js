import React, { Component } from 'react';
import PropTypes from 'prop-types';


class Searchbar extends Component {

  static propTypes = {
    changeSearchResultError: PropTypes.func.isRequired,
    addToList: PropTypes.func.isRequired
  }

  state = {
    query: '',
    videoId: ''
  }


  handleSubmit = (e) => {
    e.preventDefault();
    this.detectID();
    this.clearID();
    this.props.changeSearchResultError(false);
  }

  //sets new id and calls function that fetch info about video
  setId = (newId) => {
    this.setState({videoId: newId});
    this.clearQuery();
    this.getInfo(newId);
  }

  //fetch information about video from youtube and add item to list or set error state in App.js
  getInfo = (id) => {
    const url = `https://www.googleapis.com/youtube/v3/videos?id=${id}&key=AIzaSyB7asSzTvcMogycBslu8o4RB3DjOumaqtA&part=snippet,contentDetails,statistics,status`
    fetch(url)
      .then((response) => response.json())
      .then((result) => {
        if(result.items[0]) {
          this.props.addToList({id: result.items[0].id, favorite: false, date: `${new Date(result.items[0].snippet.publishedAt).toDateString()}`, title: result.items[0].snippet.title, views: result.items[0].statistics.viewCount, likes: result.items[0].statistics.likeCount, thumbnails: result.items[0].snippet.thumbnails.default.url });
        } else {
          this.props.changeSearchResultError(true);
        }
      })
  }

  //check what version of link is it and call function that set state
  detectID = () => {
    let query = this.state.query;
    if (query.match('https://youtu.be')) {
      let id = query.split('youtu.be/')[1];
      this.setId(id);
    } else if(query.match('https://www.youtube.com')) {
              let id = query.split('v=')[1];
              this.setId(id);
            } else {
                this.setId(query);
            };
  }

  //update state
  updateQuery = (value) => {
    this.setState({query: value});
  }

  //clear query and update state
  clearQuery = () => {
    this.setState({query: ''});
  }

  //set empty id
  clearID = () => {
    this.setState({videoId: ''});
  }


  render () {

    return (
      <div className="searchbar">
        <form onSubmit={(event) => this.handleSubmit(event)}>
          Enter video URL or identifier: <br/>
          <input tabIndex='0' className="search" type="text"
            placeholder= "Let's find a video"
            value={this.state.query}
            onChange={(event) => this.updateQuery(event.target.value)}
          />
          <button tabIndex='0' className="addToList" type="submit"><span className="glyphicon glyphicon-plus"></span></button>
        </form>
      </div>
    )
  }
}

export default Searchbar;
