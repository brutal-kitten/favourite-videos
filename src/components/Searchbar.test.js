import React from 'react';
import ReactDOM from 'react-dom';
import Searchbar from './Searchbar.js';
import TestRenderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

it('renders properly', () => {
  const tree = TestRenderer.create(
    <Searchbar />).toJSON();
  expect(tree).toMatchSnapshot();
});

const mockChangeSearchResultError = jest.fn();
const mockAddToList = jest.fn();

const props = {
  changeSearchResultError: mockChangeSearchResultError,
  addToList: mockAddToList
}

const initialState = {
  query: '',
  videoId: ''
}

it('initialize Searchbar with correct initial state', () => {
  const wrapper = shallow(<Searchbar {...props}/>);
  expect(wrapper.state()).toEqual(initialState);
});

describe('methods that set state works proprerly', () => {
  const wrapper = shallow(<Searchbar {...props}/>);

  it('should update query every time user types into input form', () => {
    wrapper.find('input').simulate('change', {
      target: {value: 'lalala'},
      preventDefault: () => {},
    });
    expect(wrapper.state('query')).toBe('lalala');
  });

  it('clearQuery method should set query to an empty string', () => {
    const instance = wrapper.instance();
    instance.clearQuery();

    expect(wrapper.state('query')).toBe('');
  });

  it('updateVideoId should set videoId with a given value', () => {
    const instance = wrapper.instance();
    instance.updateVideoId('lalala');

    expect(wrapper.state('videoId')).toBe('lalala');
  })

  it('clearID method should set videoId to an empty string', () => {
    const instance = wrapper.instance();
    instance.clearID();
    expect(wrapper.state('videoId')).toBe('');
  });
});


describe('detectID method works correctly', () => {
  const wrapper = shallow(<Searchbar {...props}/>);
  const inputs = ["WEkSYw3o5is", "https://www.youtube.com/watch?v=WEkSYw3o5is", "https://youtu.be/WEkSYw3o5is"];

  it('should detect ID from string', () => {
    wrapper.setState({query: inputs[0]});
    const instance = wrapper.instance();
    instance.detectID();
    expect(wrapper.state('videoId')).toBe("WEkSYw3o5is");
  });

  it('should detect ID from youtube link', () => {
    wrapper.setState({query: inputs[1]});
    const instance = wrapper.instance();
    instance.detectID();
    expect(wrapper.state('videoId')).toBe("WEkSYw3o5is");
  });

  it('should detect ID from youtube share link', () => {
    wrapper.setState({query: inputs[2]});
    const instance = wrapper.instance();
    instance.detectID();
    expect(wrapper.state('videoId')).toBe("WEkSYw3o5is");
  });
});

describe('submit buttons work correctly', () => {
  const wrapper = shallow(<Searchbar {...props}/>);

  const instance = wrapper.instance();
  const detectIDSpy = jest.spyOn(instance, 'detectID');
  const clearIDSpy = jest.spyOn(instance, 'clearID');
  const setIdSpy = jest.spyOn(instance, 'setId');
  const updateVideoIdSpy = jest.spyOn(instance, 'updateVideoId');
  const clearQuerySpy = jest.spyOn(instance, 'clearQuery');
  const getInfoSpy = jest.spyOn(instance, 'getInfo');

  wrapper.find('input').simulate('change', {
    target: {value: 'lalala'},
    preventDefault: () => {},
  });

  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });

  it('should call detectID', () => {
    expect(detectIDSpy).toHaveBeenCalled();
    detectIDSpy.mockRestore();
  });

  it('should call setIdSpy', () => {
    expect(setIdSpy).toHaveBeenCalled();
    setIdSpy.mockRestore();
  });

  it('should call updateVideoIdSpy', () => {
    expect(updateVideoIdSpy).toHaveBeenCalled();
    updateVideoIdSpy.mockRestore();
  });

  it('should call clearQuerySpy', () => {
    expect(clearQuerySpy).toHaveBeenCalled();
    clearQuerySpy.mockRestore();
  });

  it('should call getInfoSpy', () => {
    expect(getInfoSpy).toHaveBeenCalled();
    getInfoSpy.mockRestore();
  });

  it('should call clearIDSpy', () => {
    expect(clearIDSpy).toHaveBeenCalled();
    clearIDSpy.mockRestore();
  });

  it('should call changeSearchResultError once', () => {
    expect(mockChangeSearchResultError.mock.calls.length).toBe(1);
    expect(mockChangeSearchResultError.mock.calls[0][0]).toBe(false);
  });

});

describe('handleResult should work correct', () => {
  const wrapper = shallow(<Searchbar {...props}/>);
  const instance = wrapper.instance();

  it('should call addToList ones with  created object', () => {
    const instance = wrapper.instance();
    instance.handleResult({
    "items": [
      {
        "id": "hY7m5jjJ9mM",
        "snippet": {
          "publishedAt": "2017-05-31T09:30:02.000Z",
          "channelId": "UC9obdDRxQkmn_4YpcBMTYLw",
          "title": "CATS will make you LAUGH YOUR HEAD OFF - Funny CAT compilation",
          "thumbnails": {
            "default": {
              "url": "https://i.ytimg.com/vi/hY7m5jjJ9mM/default.jpg",
              "width": 120,
              "height": 90
            }
          },
        },
        "statistics": {
          "viewCount": "82754092",
          "likeCount": "333383",
        }
      }
    ]
  });

    expect(mockAddToList.mock.calls.length).toBe(1);
    expect(mockChangeSearchResultError.mock.calls.length).toBe(1);
    expect(mockAddToList.mock.calls[0][0]).toEqual({"date": "Wed May 31 2017", "favorite": false, "id": "hY7m5jjJ9mM", "likes": "333383", "thumbnails": "https://i.ytimg.com/vi/hY7m5jjJ9mM/default.jpg", "title": "CATS will make you LAUGH YOUR HEAD OFF - Funny CAT compilation", "views": "82754092"});
  });

  it('should changeSearchResultError with true if there is no result', () => {
    const instance = wrapper.instance();
    instance.handleResult({"items": []});
    expect(mockChangeSearchResultError.mock.calls.length).toBe(2);
    expect(mockChangeSearchResultError.mock.calls[1][0]).toBe(true);
  });
});

describe('in getInfo catch work properly', async() => {
  const wrapper = shallow(<Searchbar {...props}/>);

  it('should catch error', async () => {
    const instance = wrapper.instance();
    window.fetch = jest.fn().mockImplementation(() => {
    return new Promise((resolve, reject) => {
      reject({error : "error"});
    })});
    let response = await instance.getInfo("hY7m5jjJ9mM");
    expect(mockChangeSearchResultError.mock.calls.length).toBe(2);
    expect(mockChangeSearchResultError.mock.calls[1][0]).toBe(true);
  });
})
