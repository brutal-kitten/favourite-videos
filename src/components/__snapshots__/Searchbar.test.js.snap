// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div
  className="searchbar"
>
  <form
    onSubmit={[Function]}
  >
    Enter video URL or identifier: 
    <br />
    <input
      className="search"
      onChange={[Function]}
      placeholder="Let's find a video"
      tabIndex="0"
      type="text"
      value=""
    />
    <button
      className="addToList"
      tabIndex="0"
      type="submit"
    >
      <span
        className="glyphicon glyphicon-plus"
      />
    </button>
  </form>
</div>
`;
